x-pack/plugins/enterprise_search/public/applications/app_search/components/schema/schema_logic.ts

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+import { isEqual } from 'lodash';
+
+import {
+  flashAPIErrors,
+  setErrorMessage,
+  flashSuccessToast,
+  clearFlashMessages,
+} from '../../../shared/flash_messages';
+import { HttpLogic } from '../../../shared/http';
+import { Schema, SchemaType, IndexJob } from '../../../shared/schema/types';
+import { EngineLogic } from '../engine';
+
+import { ADD_SCHEMA_ERROR, ADD_SCHEMA_SUCCESS, UPDATE_SCHEMA_SUCCESS } from './constants';
+import { SchemaBaseLogic, SchemaBaseValues, SchemaBaseActions } from './schema_base_logic';
+import { SchemaApiResponse } from './types';
+
+interface SchemaValues extends SchemaBaseValues {
+  isUpdating: boolean;
+  hasSchema: boolean;
+  hasSchemaChanged: boolean;
+  cachedSchema: Schema;
+  mostRecentIndexJob: Partial<IndexJob>;
+  unconfirmedFields: string[];
+  hasUnconfirmedFields: boolean;
+  hasNewUnsearchedFields: boolean;
+  isModalOpen: boolean;
+}
+
+interface SchemaActions extends SchemaBaseActions {
+  loadIndexedEngineSchema(): void;
+  onSchemaLoad(response: SchemaApiResponse): SchemaApiResponse;
+  addSchemaField(
+    fieldName: string,
+    fieldType: SchemaType
+  ): { fieldName: string; fieldType: SchemaType };
+  updateSchemaFieldType(
+    fieldName: string,
+    fieldType: SchemaType
+  ): { fieldName: string; fieldType: SchemaType };
+  updateSchema(successMessage?: string): string | undefined;
+  onSchemaUpdateError(): void;
+  openModal(): void;
+  closeModal(): void;
+}
+
+export const SchemaLogic = kea<MakeLogicType<SchemaValues, SchemaActions>>({
+  path: ['enterprise_search', 'app_search', 'schema_logic'],
+  connect: {
+    values: [SchemaBaseLogic, ['dataLoading', 'schema']],
+    actions: [SchemaBaseLogic, ['loadSchema', 'setSchema']],
+  },
```

Comment:
FWIW though, I do actually kind of like that shared values/actions are so explicitly listed rather than `.extend()` being _somewhat_ magical / having to dive into another file to see what's shared. Tradeoffs everywhere I guess!